File: benchmarks/wrappers/local/python/storage.py
import os
import uuid

import minio


class storage:
    instance = None
    client = None

    def __init__(self):
        if "MINIO_ADDRESS" in os.environ:
            address = os.environ["MINIO_ADDRESS"]
            access_key = os.environ["MINIO_ACCESS_KEY"]
            secret_key = os.environ["MINIO_SECRET_KEY"]
            self.client = minio.Minio(
                address, access_key=access_key, secret_key=secret_key, secure=False
            )

    @staticmethod
    def unique_name(name):
        name, extension = os.path.splitext(name)
        return "{name}.{random}{extension}".format(
            name=name, extension=extension, random=str(uuid.uuid4()).split("-")[0]
        )

    def upload(self, bucket, file, filepath, unique_name=True):
        key_name = storage.unique_name(file) if unique_name else file
        self.client.fput_object(bucket, key_name, filepath)
        return key_name

    def download(self, bucket, file, filepath):
        data = self.client.get_object(bucket, file)
        size = data.headers.get("Content-Length")
        if size:
            os.environ["STORAGE_DOWNLOAD_BYTES"] = str(
                int(os.getenv("STORAGE_DOWNLOAD_BYTES", "0")) + int(size)
            )
        self.client.fget_object(bucket, file, filepath)

    def download_directory(self, bucket, prefix, path):
        objects = self.client.list_objects_v2(bucket, prefix, recursive=True)
        for obj in objects:
            file_name = obj.object_name
            self.download(bucket, file_name, os.path.join(path, file_name))

    def upload_stream(self, bucket, file, bytes_data):
        key_name = storage.unique_name(file)
        self.client.put_object(bucket, key_name, bytes_data, bytes_data.getbuffer().nbytes)
        return key_name

    def download_stream(self, bucket, file):
        data = self.client.get_object(bucket, file)
        body = data.read()
        os.environ["STORAGE_DOWNLOAD_BYTES"] = str(
            int(os.getenv("STORAGE_DOWNLOAD_BYTES", "0")) + len(body)
        )
        return body

    def download_within_range(self, bucket, file, start_byte, stop_byte):
        range_header = f"bytes={start_byte}-{stop_byte}"
        resp = self.client.get_object(bucket, file, request_headers={"Range": range_header})
        data = resp.read().decode("utf-8")
        os.environ["STORAGE_DOWNLOAD_BYTES"] = str(
            int(os.getenv("STORAGE_DOWNLOAD_BYTES", "0")) + len(data.encode("utf-8"))
        )
        return data

    def list_directory(self, bucket, prefix):
        if hasattr(self.client, "list_objects_v2"):
            iterator = self.client.list_objects_v2(bucket, prefix, recursive=True)
        else:
            iterator = self.client.list_objects(bucket, prefix, recursive=True)
        for obj in iterator:
            yield obj.object_name

    def get_instance():
        if storage.instance is None:
            storage.instance = storage()
        return storage.instance

File: benchmarks/wrappers/local/golang/storage.go
package golang

import (
    "bytes"
    "context"
    "fmt"
    "io"
    "os"
    "path/filepath"
    "strconv"
    "strings"

    "github.com/google/uuid"
    "github.com/minio/minio-go/v7"
    "github.com/minio/minio-go/v7/pkg/credentials"
)

type Storage struct {
    client *minio.Client
}

var instance *Storage

func UniqueName(name string) string {
    ext := filepath.Ext(name)
    basename := name[:len(name)-len(ext)]
    uuidStr := strings.Split(uuid.New().String(), "-")[0]
    return fmt.Sprintf("%s.%s%s", basename, uuidStr, ext)
}

func NewStorage() (*Storage, error) {
    address, exists := os.LookupEnv("MINIO_ADDRESS")
    if !exists {
        return nil, nil
    }
    
    accessKey := os.Getenv("MINIO_ACCESS_KEY")
    secretKey := os.Getenv("MINIO_SECRET_KEY")
    
    parts := strings.Split(address, ":")
    endpoint := parts[0]
    var port int = 9000 // Default port
    
    if len(parts) > 1 {
        var err error
        port, err = strconv.Atoi(parts[1])
        if err != nil {
            return nil, fmt.Errorf("invalid port in MINIO_ADDRESS: %v", err)
        }
    }

    client, err := minio.New(endpoint, &minio.Options{
        Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
        Secure: false,
        Port:   uint16(port),
    })
    if err != nil {
        return nil, err
    }

    return &Storage{client: client}, nil
}

func (s *Storage) Upload(bucket, file, filepath string) (string, error) {
    keyName := UniqueName(file)
    _, err := s.client.FPutObject(context.Background(), bucket, keyName, filepath, minio.PutObjectOptions{})
    if err != nil {
        return "", err
    }
    return keyName, nil
}

func (s *Storage) Download(bucket, file, filepath string) error {
    return s.client.FGetObject(context.Background(), bucket, file, filepath, minio.GetObjectOptions{})
}

func (s *Storage) DownloadDirectory(bucket, prefix, path string) error {
    ctx := context.Background()
    objects := s.client.ListObjects(ctx, bucket, minio.ListObjectsOptions{
        Prefix:    prefix,
        Recursive: true,
    })

    for object := range objects {
        if object.Err != nil {
            return object.Err
        }
        
        objectPath := filepath.Join(path, object.Key)
        if err := os.MkdirAll(filepath.Dir(objectPath), 0755); err != nil {
            return err
        }
        
        if err := s.Download(bucket, object.Key, objectPath); err != nil {
            return err
        }
    }
    
    return nil
}

func (s *Storage) UploadStream(bucket, file string, data []byte) (string, error) {
    keyName := UniqueName(file)
    reader := bytes.NewReader(data)
    _, err := s.client.PutObject(context.Background(), bucket, keyName, reader, int64(len(data)), minio.PutObjectOptions{})
    if err != nil {
        return "", err
    }
    return keyName, nil
}

func (s *Storage) DownloadStream(bucket, file string) ([]byte, error) {
    obj, err := s.client.GetObject(context.Background(), bucket, file, minio.GetObjectOptions{})
    if err != nil {
        return nil, err
    }
    defer obj.Close()
    
    return io.ReadAll(obj)
}

func GetInstance() (*Storage, error) {
    if instance == nil {
        var err error
        instance, err = NewStorage()
        if err != nil {
            return nil, err
        }
    }
    return instance, nil
}